pkg/agent/nodeportlocal/portcache/port_table.go

```diff
@@ -64,94 +93,134 @@ func NewPortTable(start, end int) (*PortTable, error) {
 func (pt *PortTable) CleanupAllEntries() {
 	pt.tableLock.Lock()
 	defer pt.tableLock.Unlock()
-	pt.Table = make(map[int]NodePortData)
+	pt.Table = make(map[int]*NodePortData)
+	pt.PodTable = make(map[string]*NodePortData)
 }
 
 func (pt *PortTable) GetEntry(nodeport int) *NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
 	data, _ := pt.Table[nodeport]
-	return &data
+	return data
 }
 
 func (pt *PortTable) GetDataForPodIP(ip string) []NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
 	var allData []NodePortData
-	for _, data := range pt.Table {
-		if data.PodIP == ip {
-			allData = append(allData, data)
+	for i := range pt.Table {
+		if (*pt.Table[i]).PodIP == ip {
+			allData = append(allData, *pt.Table[i])
 		}
 	}
 	return allData
 }
 
-func (pt *PortTable) GetEntryByPodIPPort(ip string, port int) *NodePortData {
+func (pt *PortTable) GetEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
-	return pt.getEntryByPodIPPort(ip, port)
+	return pt.getEntryByPodIPPortProtocol(ip, port, protocol)
 }
```

Comment:
this should return a copy of `NodePortData` or there could be some thread safety issues?

Reply:
I moved this into GetEntry itself. I have also kept it to return a pointer to `NodePortData`. `GetEntry` is called from `RuleExists` and does a nil check. Other functions that expect `*NodePortData` have table locks implemented, so it should be fine?